riscv/triggers.cc

```diff
@@ -148,6 +148,59 @@ match_result_t mcontrol_t::memory_access_match(processor_t * const proc, operati
   return MATCH_NONE;
 }
 
+reg_t itrigger_t::tdata1_read(const processor_t * const proc) const noexcept
+{
+  auto xlen = proc->get_xlen();
+  reg_t tdata1 = 0;
+  tdata1 = set_field(tdata1, CSR_ITRIGGER_TYPE(xlen), CSR_TDATA1_TYPE_ITRIGGER);
+  tdata1 = set_field(tdata1, CSR_ITRIGGER_DMODE(xlen), dmode);
+  tdata1 = set_field(tdata1, CSR_ITRIGGER_HIT(xlen), hit);
+  tdata1 = set_field(tdata1, CSR_ITRIGGER_VS, vs);
+  tdata1 = set_field(tdata1, CSR_ITRIGGER_VU, vu);
+  tdata1 = set_field(tdata1, CSR_ITRIGGER_NMI, nmi);
+  tdata1 = set_field(tdata1, CSR_ITRIGGER_M, m);
+  tdata1 = set_field(tdata1, CSR_ITRIGGER_S, s);
+  tdata1 = set_field(tdata1, CSR_ITRIGGER_U, u);
+  tdata1 = set_field(tdata1, CSR_ITRIGGER_ACTION, action);
+  return tdata1;
+}
+
+void itrigger_t::tdata1_write(processor_t * const proc, const reg_t val) noexcept
+{
+  auto xlen = proc->get_xlen();
+  assert(get_field(val, CSR_ITRIGGER_TYPE(xlen)) == 4);
+  dmode = proc->get_state()->debug_mode ? get_field(val, CSR_ITRIGGER_DMODE(xlen)) : 0;
+  hit = get_field(val, CSR_ITRIGGER_HIT(xlen));
+  vs = proc->extension_enabled('H') ? get_field(val, CSR_ITRIGGER_VS) : 0;
+  vu = proc->extension_enabled('H') ? get_field(val, CSR_ITRIGGER_VU) : 0;
+  nmi = get_field(val, CSR_ITRIGGER_NMI);
+  m = get_field(val, CSR_ITRIGGER_M);
+  s = proc->extension_enabled('S') ? get_field(val, CSR_ITRIGGER_S) : 0;
+  u = proc->extension_enabled('U') ? get_field(val, CSR_ITRIGGER_U) : 0;
+  action = (action_t)get_field(val, CSR_ITRIGGER_ACTION);
+  if (action > 4 || (action==ACTION_DEBUG_MODE && dmode==0))
+    action = ACTION_DEBUG_EXCEPTION;
+}
+
+match_result_t itrigger_t::trap_taking_match(processor_t * const proc, trap_t& t)
```

Comment:
Can this be `const trap_t& t`?

Reply:
This requires trap_t::cause() to be a const function. I add a refactoring commit to do so.